# Analysis of single cell RNAseq (scRNAseq) data

## Description

**NOTE: this material is currently under development**

[[links to other materials that have been adapted in this lesson]]

Software requirements for this course can be found on [fredhutch.io's Software page](http://www.fredhutch.io/software/).

## Tutorials

- [cellranger.md](cellranger.md): fastq file processing and mapping using on-premise compute resources at Fred Hutch.
Requires an understanding of command line operations using [Unix/bash shell](https://github.com/fredhutchio/unix_shell_intro).
- [monocle](monocle/): analysis and visualization of processed scRNAseq data using `monocle` and `garnett`,
which requires an understanding of [R code](https://github.com/fredhutchio/r_intro).

## Resources

* Materials are described in the relevant markdown (`*.md`) files.
* [`solutions/`](solutions/) includes the solutions for all challenge questions included in tutorials.
* [`instructors.md`](instructors.md) includes information for instructors to facilitate teaching each lesson
* [`hackmdio.txt`](hackmdio.txt) is an archive of the [interactive webpage](https://hackmd.io) used during lessons
